python/ray/includes/serialization.pxi

```diff
@@ -0,0 +1,337 @@
+        buffer.set_itemsize(view.itemsize)
+        if view.format:
+            buffer.set_format(view.format)
+        if view.shape:
+            for i in range(view.ndim):
+                buffer.add_shape(view.shape[i])
+        if view.strides:
+            for i in range(view.ndim):
+                buffer.add_strides(view.strides[i])
+
+        # Increase buffer address.
+        if view.len < kMajorBufferSize:
+            self._curr_buffer_addr = padded_length(
+                self._curr_buffer_addr, kMinorBufferAlign)
+        else:
+            self._curr_buffer_addr = padded_length(
+                self._curr_buffer_addr, kMajorBufferAlign)
+        buffer.set_address(self._curr_buffer_addr)
+        self._curr_buffer_addr += view.len
+        self.buffers.push_back(view)
+
+    def get_total_bytes(self, const c_string &inband):
+        cdef:
+            size_t protobuf_bytes = 0
+            uint64_t buffers_offset
+            uint64_t inband_data_offset = sizeof(int64_t) * 2
+        buffers_offset = padded_length_u64(
+            inband_data_offset + inband.length(), kMajorBufferAlign)
+        self._protobuf_offset = padded_length_u64(
+            buffers_offset + self._curr_buffer_addr, kMinorBufferAlign)
+        # This is a placeholder for correct calculation of size.
+        self.python_object.set_inband_data_offset(inband_data_offset)
+        self.python_object.set_inband_data_size(inband.length())
+        self.python_object.set_buffers_offset(buffers_offset)
+        # We MUST NOT change 'python_object' size since now.
+        protobuf_bytes = self.python_object.ByteSizeLong()
+        if protobuf_bytes > INT32_MAX:
+            raise ValueError("Total buffer metadata size is bigger than %d. "
+                             "Consider reduce the number of buffers "
+                             "(number of numpy arrays, etc)." % INT32_MAX)
+        self._total_bytes = self._protobuf_offset + protobuf_bytes
+        return self._total_bytes
+
+    cdef void write_to(self, const c_string &inband, shared_ptr[CBuffer] data,
+                       int memcopy_threads):
+        cdef uint8_t *ptr = data.get().Data()
+        cdef uint8_t *__temp
+        cdef int32_t protobuf_size
+        cdef uint64_t buffer_addr
+        cdef uint64_t buffer_len
+        cdef int i
+        if self._total_bytes < 0:
+            raise ValueError("Must call 'get_total_bytes()' first "
+                             "to get the actual size")
+        # Write protobuf size for deserialization.
+        protobuf_size = self.python_object.GetCachedSize()
+        (<int64_t*>ptr)[0] = self._protobuf_offset
+        (<int64_t*>ptr)[1] = protobuf_size
+        # Write protobuf data.
+        self.python_object.SerializeWithCachedSizesToArray(
+            ptr + self._protobuf_offset)
+        # Write inband data.
+        memcpy(ptr + self.python_object.inband_data_offset(),
+               inband.c_str(), inband.length())
```

Comment:
Is there a reason you didn't call `inband.data()`? It doesn't need to be null terminated, right?

Reply:
oh, that's true. I should call `inband.data()`. thanks for mention it.